WalletWasabi.Tests/ExtPubKeyExplorerTests.cs

```diff
@@ -0,0 +1,158 @@
+			};
+
+			var explorer = new ExtPubKeyExplorer(_extPubKey, filters);
+
+			var unusedKeyIndex = explorer.GetIndexFirstUnusedKey();
+			Assert.Equal(0, unusedKeyIndex);
+		}
+
+		[Fact]
+		public void ShouldFindUnsedKeyFirstChunk()
+		{
+			var filters = new []{
+				CreateFiltersWith(GetScripts(0, 10)),
+				CreateFiltersWith(GetScripts(10, 10))
+			};
+
+			var explorer = new ExtPubKeyExplorer(_extPubKey, filters);
+
+			var unusedKeyIndex = explorer.GetIndexFirstUnusedKey();
+			Assert.Equal(20, unusedKeyIndex);
+		}
+
+		[Fact]
+		public void ShouldFindUnsedKeyAtTheEndOfFirstChunk()
+		{
+			var filters = new []{
+				CreateFiltersWith(GetScripts(0, 500)),
+				CreateFiltersWith(GetScripts(500, 499))
+			};
+
+			var explorer = new ExtPubKeyExplorer(_extPubKey, filters);
+
+			var unusedKeyIndex = explorer.GetIndexFirstUnusedKey();
+			Assert.Equal(999, unusedKeyIndex);
+		}
+
+		[Fact]
+		public void ShouldFindUnsedKeyAnywhereFirstChunk()
+		{
+			var filters = new []{
+				CreateFiltersWith(GetScripts(0,  1)),
+				CreateFiltersWith(GetScripts(0,  2)),
+				CreateFiltersWith(GetScripts(0, 27)),
+				CreateFiltersWith(GetScripts(27, 3)),
+			};
+
+			var explorer = new ExtPubKeyExplorer(_extPubKey, filters);
+
+			var unusedKeyIndex = explorer.GetIndexFirstUnusedKey();
+			Assert.Equal(30, unusedKeyIndex);
+		}
+
+		[Fact]
+		public void ShouldFindUnsedKeySecondChunk()
+		{
+			var filters = new []{
+				CreateFiltersWith(GetScripts(  0, 250)),
+				CreateFiltersWith(GetScripts(250, 250)),
+				CreateFiltersWith(GetScripts(500, 250)),
+				CreateFiltersWith(GetScripts(750, 250))
+			};
+
+			var explorer = new ExtPubKeyExplorer(_extPubKey, filters);
+
+			var unusedKeyIndex = explorer.GetIndexFirstUnusedKey();
+			Assert.Equal(1_000, unusedKeyIndex);
+		}
+
+		[Fact]
+		public void ShouldFindUnsedKeyFarFarAway()
+		{
+			var filters = Enumerable.Range(0, 100).Select(x=>CreateFiltersWith(GetScripts(x*250, 250))).ToArray();
+
+			var explorer = new ExtPubKeyExplorer(_extPubKey, filters);
+
+			var unusedKeyIndex = explorer.GetIndexFirstUnusedKey();
+			Assert.Equal(25_000, unusedKeyIndex);
+		}
+
+		[Fact]
+		public void RealWorldCase()
+		{
+			var filters = new List<FilterModel>(100_000);
+			using(var file = File.OpenRead("../../../../WalletWasabi.Bench/IndexTestNet.bin"))
+			while(file.Position < file.Length)
+				filters.Add(FilterModel.FromStream(file, 0));
+
+			var extPubKey = ExtPubKey.Parse("tpubDESeudcLbEHBoH8iw6mL284PXoe2TVmGa3MdQ2gSAkkMj9d1P88LB8wEbVYpigwzurDmDSRsGMKkUsH6vx1anBDCMRzha4YucJfCvEy6z6B");
+			var explorer = new ExtPubKeyExplorer(extPubKey, filters, 3); 
+			var index = explorer.GetIndexFirstUnusedKey();
+			Console.WriteLine($"first index: m/84'/0'/0'/0/{index}");
+			// Review: The Wallet file indicates that the first Locked index is 48, however increasing the chunkSize makes the
```

Comment:
I need another pair of eyes (and another brain) here. For some reason that I don't understand if I increase the number of scriptPubKeys to test in each iteration the algorithm returns a wrong index value. This issue was driving me crazy all the day. Everything works okay a pretty fast except for this problem. @molnard could you give it a look, please?

Reply:
I solved the mystery while sleeping. Our filters have a false positive rate (Fpr) `1/2^20` or about `0.000001`, that means that if we try with lets say 1,000 addresses at a time the probability of getting a Fp in any given filter is `0.001` so, given we have almost 100,000 non-empty filters in testnet the **probability of getting a false positive is 1**`. 

My idea of trying with big sets of addresses to speed up the discovery is now clearly wrong. However the algorithm can be used in combination with a double-checking mechanism such that when the `Match` method returns true it can be double-checked try one by one. 

I will implement that change now.